flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/BufferedUpsertWriter.java

```diff
@@ -0,0 +1,209 @@
+    private final ScheduledFuture<?> scheduledFuture;
+    private final int batchMaxRowNums;
+    private final Function<RowData, RowData> valueCopier;
+    private final Map<RowData, Tuple2<RowData, Long>> reduceBuffer = new HashMap<>();
+    private final Function<RowData, RowData> keyExtractor;
+
+    private boolean closed = false;
+    private int batchCount = 0;
+    private volatile Exception flushException;
+
+    BufferedUpsertWriter(
+            SinkWriter<RowData, ?, WriterState> wrappedWriter,
+            DataType physicalDataType,
+            int[] keyProjection,
+            SinkBufferFlushMode bufferFlushMode,
+            @Nullable TypeSerializer<RowData> typeSerializer) {
+        checkArgument(bufferFlushMode != null && bufferFlushMode.isEnabled());
+        this.wrappedWriter = checkNotNull(wrappedWriter);
+        this.batchMaxRowNums = bufferFlushMode.getBatchSize();
+        long batchIntervalMs = bufferFlushMode.getBatchIntervalMs();
+
+        List<LogicalType> fields = physicalDataType.getLogicalType().getChildren();
+        final RowData.FieldGetter[] keyFieldGetters =
+                Arrays.stream(keyProjection)
+                        .mapToObj(
+                                targetField ->
+                                        RowData.createFieldGetter(
+                                                fields.get(targetField), targetField))
+                        .toArray(RowData.FieldGetter[]::new);
+        this.keyExtractor = rowData -> createProjectedRow(rowData, RowKind.INSERT, keyFieldGetters);
+        this.valueCopier = typeSerializer == null ? Function.identity() : typeSerializer::copy;
+
+        this.scheduledFuture =
+                this.scheduler.scheduleWithFixedDelay(
+                        () -> {
+                            synchronized (BufferedUpsertWriter.this) {
+                                if (!closed) {
+                                    try {
+                                        flush();
+                                    } catch (Exception e) {
+                                        flushException = e;
+                                    }
+                                }
+                            }
+                        },
+                        batchIntervalMs,
+                        batchIntervalMs,
+                        TimeUnit.MILLISECONDS);
+    }
+
+    @Override
+    public void write(RowData element, Context context) throws IOException, InterruptedException {
+        wrappedContext.setContext(context);
+        addToBuffer(element, context.timestamp());
+    }
+
+    @Override
+    public List<Void> prepareCommit(boolean flush) throws IOException, InterruptedException {
+        flush();
+        return Collections.emptyList();
+    }
+
+    @Override
+    public List<WriterState> snapshotState() throws IOException {
+        return wrappedWriter.snapshotState();
```

Comment:
We need to snapshot the buffer as well. Or else we do not even get at least once.

Reply:
I'll create a ticket as follow-up once this PR is merged to fix in the new sink and also fix in the old one.